# Pandas: Data Management 1. Importing, segmenting, and summarizing data. 

*Authors: Evan Carey and Jemma Nelson, written for BH Analytics*

## Overview

In this section, we will go over the introductory aspects of the Pandas package. This is the primary package we will use for data management when doing data analysis in Python. The Pandas module provides dataframe object, which is the 'spreadsheet' like data object in Python. The Pandas module also provides the series object, which is a 1-dimensional array (one column of a dataframe). 

Our Objectives for this section are:

*  Import data with Pandas
*  Understand Dataframes and Series object types
*  Subset and manipulate dataframes

## Data for this Session: Healthcare Visits

To demonstrate these concepts, we will use some simulated data from a health care system. There are two files we will use throughout this section. 

*  The first file is called `Patient.csv`, and is information about patients in the healthcare system. There should be one row per patient in this file, so we call this a patient level file. 

*  The second file is called `OutpatientVisit.csv`, and is information about individual visits to the doctor for patients in the healthcare system. There will be more than one row per patient in this file, since patients can have more than one visit. However, there should only be one row per visit, so we call this a visit-level file. 

The files are located here: 

* Data/Data_Sims/healthcare/Patient.csv
* Data/Data_Sims/healthcare/OutpatientVisit.csv

## Libraries

In [1]:
import sys
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import os
import textwrap

In [2]:
# Enable inline plotting for graphics
%matplotlib inline

There are a few new packages here we are calling! 

* Pandas is the data management package that we will focus on. 
* Numpy is the numerical computation package in Python. You can think of this as part of the engine under the hood of Pandas...
* Matplotlib is the main plotting package in Python
* Seaborn is an 'add-on' plotting package that is based on Matplotlib. More details on these to come later!

In [3]:
# Get Version information
print(textwrap.fill(sys.version),'\n')
print("Pandas version: {0}".format(pd.__version__),'\n')
print("Matplotlib version: {0}".format(matplotlib.__version__),'\n')
print("Numpy version: {0}".format(np.__version__),'\n')
print("Seaborn version: {0}".format(sns.__version__),'\n')

3.7.10 (default, Feb 26 2021, 13:06:18) [MSC v.1916 64 bit (AMD64)] 

Pandas version: 1.2.4 

Matplotlib version: 3.3.4 

Numpy version: 1.20.1 

Seaborn version: 0.11.1 



In [4]:
# So all output comes through from Ipython
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Check your working directory

Subsequent sessions may require you to identify and update your working directory so paths correctly point at the downloaded data files. You can check your working directory like so:

In [5]:
# Working Directory
print(os.getcwd())

C:\Users\evancarey\Dropbox\Work\BHAnalytics\courseware\Python_Courses\course_downloads\CDC_Python_Week1\notebooks


You can simply assign the working directory to be 'one level above the current directory' by running this line of code with the `..` as the path. 
  
`..` means 'one directory above the current directory.
  
Note you should only run it once! If you run it multiple times, you will keep reseting the working directory to one level above the current directory, until you reach the root of the drive!

In [6]:
# Set Working Directory (if needed)
# you can just do this:
os.chdir(r"..") 

For the purpose of compiling these notebooks, I put my working directory here, but this line of code will give you an error since this location does not exist on your file system! I have included the `try` command to avoid the error preventing the rest of the notebook from running in case you miss this.

You could update this string: `r"C:\Users\evancarey\Dropbox\Work"`  

to be a valid path on your filesystem pointing at the course download. 

In [7]:
# Set New Working Directory 
# the try except tells Python to try to run the code
# and if an error occurs, print a message
try:
    os.chdir(r"C:\Users\evancarey\Dropbox\Work")
except:
    print("ERROR: You probably forgot to change the directory path!")

In [8]:
# Confirm it changed the working Directory
print("My working directory:\n" + os.getcwd())

My working directory:
C:\Users\evancarey\Dropbox\Work


## Set options

Here I set the max printing rows to be 10, so I don't overwhelm the printed workbooks or these presentation materials. You can change this to a larger number since you are running this on your own machine and not presenting!

In [9]:
# increase this beyond 10 if you want!
pd.options.display.max_rows = 10

## Importing structured data

We generally start by examining and importing a dataset. In practice, you will have been given a dataset that you will use Python to further analyze. This data may come in many formats...delimited text files, structured text files, Excel files, SQL database, SAS files, etc...

> You should always preserve your source data in it's original format. Don't ever change your source data! 

The first step is to examine the dataset before you import it to Python (if possible). 
* For a text file (like a CSV), that means opening it in a text editor. Do not use Excel to open a CSV, use an actual text editor. 

We will focus on delimited text in this course, but Pandas (and Python in general) has the ability to import from a wide variety of data sources, including SQL connections. 


We can import structured data using the `pd.read_csv()` function, or more generally using the `pd.read_table()` function. Check the help for both of these functions to see the different arguments.

In [10]:
#?pd.read_csv
#?pd.read_table

We will start by importing the patient and visits data. Check both of these files using a text editor first:

* Do they have column names in the first row?
* What is the delimiter? Is it actually a CSV?
* How many rows and columns does the file have? 

In [11]:
## import data
df_patient = \
    pd.read_csv("Data/Data_Sims/healthcare/Patient.csv")
df_visits = \
    pd.read_csv("Data/Data_Sims/healthcare/OutpatientVisit.csv")

Following import, you can examine the top and bottom of the dataframe simply by calling the object:

In [12]:
df_patient

,PatientID,FirstName,LastName,State,ZipCode,DateOfBirth,Gender,Race,Income
0,1,Loretta,Gunter,FL,32250,1979-03-29,female,white,29.0
1,2,Todd,Rea,TX,79602,1936-12-20,male,Missing,NaN
2,3,Margaret,Goodwin,PA,18106,1948-04-19,female,hispanic,53.0
3,4,Anna,McCullough,TX,75039,1997-08-28,female,Missing,20.0
4,5,Glenn,Labrecque,NM,87102,1985-08-19,male,Missing,48.0
...,...,...,...,...,...,...,...,...,...
19995,19996,Lucia,Atkins,MI,49546,1970-01-26,female,white,166.0
19996,19997,Wilfredo,Reinhardt,CO,80112,1950-02-23,male,other,31.0
19997,19998,Thanh,Large,FL,33912,1957-11-24,male,black,60.0
19998,19999,Deidre,Croft,GA,30303,1965-01-24,female,Unknown,NaN


In [13]:
df_visits

,VisitID,StaffID,PatientID,VisitDate,ICD10_1,ICD10_2,ICD10_3,ClinicCode
0,1,24,1,2011-08-05,G801,NaN,NaN,7
1,2,13,1,2013-06-15,G801,NaN,NaN,49
2,3,36,1,2013-12-28,G801,NaN,NaN,42
3,4,14,1,2014-10-21,G801,NaN,NaN,29
4,5,45,1,2015-05-11,G801,NaN,NaN,21
...,...,...,...,...,...,...,...,...
181391,181392,44,20000,2011-08-26,E1322,NaN,NaN,17
181392,181393,28,20000,2012-01-05,E1322,NaN,NaN,57
181393,181394,32,20000,2012-01-23,E1322,NaN,NaN,15
181394,181395,38,20000,2012-05-10,E1322,NaN,NaN,1


## Dataframe attributes

Following import, the first thing we should check is the dimensionality of the file. We want to be sure the number of rows and columns are as we expected based on our examination of the source data. 

> This is a critical but often overlooked step! Any unnoticed errors in data import may invalidate the rest of your analysis and workflow. Be careful in this step!

In [14]:
# dataframe dimensions
df_patient.shape
df_visits.shape

(20000, 9)

(181396, 8)

What type of object was returned just now? Can you describe it? This is why we must be comfortable with base Python objects...

After confirming the dimensionality, we want to examine the actual variable types (called dtypes in Pandas). This is the notion of what kind of data is contained in each column. Is it numeric (like age)? Or perhaps a string (like First Name)? We will discuss these in more detail in a moment. First, let's talk about accessing column names and row names. 

In [15]:
# List all the dtypes of the dataframe
df_patient.dtypes

PatientID        int64
FirstName       object
LastName        object
State           object
ZipCode          int64
DateOfBirth     object
Gender          object
Race            object
Income         float64
dtype: object

We can also print out a fuller data summary with the method `.info()`

In [16]:
df_patient.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PatientID    20000 non-null  int64  
 1   FirstName    20000 non-null  object 
 2   LastName     20000 non-null  object 
 3   State        20000 non-null  object 
 4   ZipCode      20000 non-null  int64  
 5   DateOfBirth  19060 non-null  object 
 6   Gender       19426 non-null  object 
 7   Race         19254 non-null  object 
 8   Income       18600 non-null  float64
dtypes: float64(1), int64(2), object(6)
memory usage: 1.4+ MB


## Series and DataFrames

We are using a **DataFrame** object from the Pandas library. 

A DataFrame is a rectangular container for data, like a spreadsheet that you are familiar with. As we've seen, it has columns and rows. Each column of a DataFrame can have its own data type-- integer, floating point, object(character), categorical, or datetime. These columns, or 1-dimension vectors of data, are called **Series.** Thus, a DataFrame is composed of any number of Series.

<table>
    <tr><th>Series1</th><th>Series2</th><th style="border:solid red;border-width:2px 2px 0px 2px;">Series3</th></tr>
    <tr><td>1</td><td>True</td><td style="border:solid red;border-width:0px 2px 0px 2px;">mouse</td></tr>
    <tr><td>2</td><td>False</td><td style="border:solid red;border-width:0px 2px 0px 2px;">cat</td></tr>
    <tr><td>3</td><td>NaN</td><td style="border:solid red;border-width:0px 2px 2px 2px;">dog</td></tr>
</table>


## Summarizing a DataFrame

As we've seen above, we can summarize a DataFrame in a number of ways. The `.dtypes` attribute or `.info()` method can give us technical information on the object.

How else do we summarize the data? In general, we want to summarize data with both numerical reports (**summary statistics**) and **visualizations.** We'll explore this more comprehensively in the *Exploratory Data Analysis* module. For now, we'll just look at quick methods.

## Numerical Summaries

For numerical summaries, we can use the `.describe()` method. 

This defaults to giving us summary statistics on numerical dtypes. We get back **measures of central tendency** (like the mean or median) and **measures of spread** (quantiles, range, standard deviation.)

Why are PatientID and ZipCode included below? When we brought in our data, Pandas assumed these were numeric data. Shortly we will correct this.

In [17]:
## Numerical Summaries

df_patient.describe()

,PatientID,ZipCode,Income
count,20000.000000,20000.000000,18600.000000
mean,10000.500000,50453.858850,116.674140
std,5773.647028,29303.181041,207.664735
min,1.000000,758.000000,3.000000
25%,5000.750000,27529.000000,34.000000
50%,10000.500000,48225.000000,59.000000
75%,15000.250000,76801.000000,104.000000
max,20000.000000,99922.000000,1780.000000


It looks like the average income in the dataset is 116.67 dollars. We could also recover that number by asking for the mean of a specific column:

In [18]:
# Use numpy function
np.mean(df_patient['Income'])
# use Series method
df_patient['Income'].mean()

116.67413978494623

116.67413978494623

How many functions like `mean()` are available? Lots! Check this link for a list of them if you are curious: We will revisit this later though :) 

https://pandas.pydata.org/pandas-docs/stable/reference/series.html#computations-descriptive-stats

Here's a table of a few useful ones:


- .mean()
- .median()
- .std()
- .min()
- .max()
- .mode()
- .sum()
- .count()

## Rows and Columns in the Dataframe

Dataframes have both row names and column names. This is a bit of a departure from other analytic languages. We often primarily think in terms of only the column names, and the row names are just incidental. 

We will see in Pandas that is not really true...row names should be thought similarly to column names (just as important). Every dataframe has row names and column names!

You can access the columns names using the `.columns` attribute. This is useful is you want to capture the column names for later use (like building a model formula, or iterating through the columns...). 

In [19]:
# List Columns
df_patient.columns

Index(['PatientID', 'FirstName', 'LastName', 'State', 'ZipCode', 'DateOfBirth',
       'Gender', 'Race', 'Income'],
      dtype='object')

The resulting object is an `Index()` object type (a Pandas thing). We generally refer to rows or columns as the index. If you want to convert this to a numpy array (a more general object), just call the `.values` attribute. 

In [20]:
df_patient.columns.values

array(['PatientID', 'FirstName', 'LastName', 'State', 'ZipCode',
       'DateOfBirth', 'Gender', 'Race', 'Income'], dtype=object)

We can change the case of the columns by iterating through the index object using string methods.

In [21]:
# Capture column names
df_patient_col_names = df_patient.columns
# Set all to lower case
df_patient.columns = [x.lower() for x in df_patient.columns]
df_patient.columns

Index(['patientid', 'firstname', 'lastname', 'state', 'zipcode', 'dateofbirth',
       'gender', 'race', 'income'],
      dtype='object')

In [22]:
# Restore originals
df_patient.columns = df_patient_col_names
df_patient.columns

Index(['PatientID', 'FirstName', 'LastName', 'State', 'ZipCode', 'DateOfBirth',
       'Gender', 'Race', 'Income'],
      dtype='object')

You can use any string method here to clean up your variable names. Here are a few ideas:


`?str.lower()`  
`?str.upper()`  
`?str.capitalize()`  
`?str.replace()`  

## Rows (index?)

I mentioned that rows are just as important as columns in Pandas. Since we did not give the dataframe any row names, they will be autogenerated as integers from 0 to the number of rows. 

You can access the rownames using `.index`. That is confusing, since I just said the word index should apply to rows and columns?!? Sorry, that is an inconsistency in Pandas. 

> The RangeIndex object returned is a generator. That just means the rownames don't quite exist yet. As soon as we need them, the RangeIndex generator will create them. 

In [23]:
df_patient.index

RangeIndex(start=0, stop=20000, step=1)

## Subsetting: Filtering by Row and Selecting Columns

After data import, our next step is to start examining and cleaning the dataframe. As part of that process, we need to able to subset the dataframe:

* You will need to be able to subset the dataframe to specific columns so you can plot them or modify them. 
* You will want to select multiple columns
* You will need to subset the dataframe by row conditions (everyone with age > 100)
* Do these all at once!

You can reference a single column like so:

In [24]:
# Indexing
# Select Column
# Returns a pandas series
df_patient['Race']

0           white
1         Missing
2        hispanic
3         Missing
4         Missing
           ...   
19995       white
19996       other
19997       black
19998     Unknown
19999     Missing
Name: Race, Length: 20000, dtype: object

This returned an object called a `Series`. 
* A Series is a single dimensional  object. 
* Dataframes are made up of a bunch of Pandas Series (each column is a Series)

Series and Dataframes do not have the same attributes and methods, so this can cause some confusion. If you are writing code and getting an error on what you think is a dataframe with only one column, check to be sure you have dataframe (instead of just a series). 

Check out this stackoverflow post for more info:

https://stackoverflow.com/questions/26047209/what-is-the-difference-between-a-pandas-series-and-a-single-column-dataframe

In [25]:
# To return a one-column data frame, pass in a list
df_patient[['Race']]

,Race
0,white
1,Missing
2,hispanic
3,Missing
4,Missing
...,...
19995,white
19996,other
19997,black
19998,Unknown


You can return multiple columns by passing in a list:

In [26]:
df_patient[['PatientID','Gender']]

,PatientID,Gender
0,1,female
1,2,male
2,3,female
3,4,female
4,5,male
...,...,...
19995,19996,female
19996,19997,male
19997,19998,male
19998,19999,female


Although accessing by attribute is widespread and convenient, note the warnings in the pandas documentation, summarized here:

> - If you try to use attribute access to create a new column, it fails silently, creating a new attribute rather than a new column.  
- The attribute will not be available if it conflicts with an existing method name


https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access

In [27]:
# By attribute
df_patient.Race # avoid

0           white
1         Missing
2        hispanic
3         Missing
4         Missing
           ...   
19995       white
19996       other
19997       black
19998     Unknown
19999     Missing
Name: Race, Length: 20000, dtype: object

After you have filtered to a column, you can apply summarizing functions. Here is a common code idiom:

In [28]:
np.mean(df_patient['Income'])

116.67413978494623

You can select by Boolean condition by passing in a Boolean series (instead of string literals) to the subset operator (the square brackets). A common idiom is to subset the data by some Boolean condition based on values in the data like so: 

In [29]:
# select by Boolean row
df_patient[df_patient['Income'] > 500]

,PatientID,FirstName,LastName,State,ZipCode,DateOfBirth,Gender,Race,Income
8,9,William,Fontenot,OH,45402,1938-04-24,male,,1093.0
60,61,Tara,Anderson,AL,36303,1879-10-30,female,white,1138.0
61,62,Diego,Yang,PA,18621,1956-11-24,male,NaN,1093.0
72,73,Lauren,Sturdevant,UT,84104,1997-04-18,female,Missing,1131.0
99,100,Amanda,Delgado,KS,67107,1976-11-05,female,,1079.0
...,...,...,...,...,...,...,...,...,...
19889,19890,Adam,Carr,LA,71101,1998-01-13,male,white,1013.0
19900,19901,James,Thurman,OH,43950,1958-10-18,male,other,1063.0
19937,19938,Curtis,Owens,CT,6492,1979-04-03,male,white,1108.0
19952,19953,Olive,Hansen,IL,60191,1960-06-28,female,white,1023.0


Only 798 people have an income above 500k per year! Notice the rownames here...do you see they were retained from the original data? Rownames matter in Pandas...(more on this later)

The square brackets following a dataframe will also accept slicing style syntax. This is the same as Numpy, and Base Python. These are end exclusive. Here is the general syntax:

`df[start:stop:increment]`

Here are a few examples:

In [30]:
# first 10 rows
df_patient[0:10]

,PatientID,FirstName,LastName,State,ZipCode,DateOfBirth,Gender,Race,Income
0,1,Loretta,Gunter,FL,32250,1979-03-29,female,white,29.0
1,2,Todd,Rea,TX,79602,1936-12-20,male,Missing,NaN
2,3,Margaret,Goodwin,PA,18106,1948-04-19,female,hispanic,53.0
3,4,Anna,McCullough,TX,75039,1997-08-28,female,Missing,20.0
4,5,Glenn,Labrecque,NM,87102,1985-08-19,male,Missing,48.0
5,6,Marlo,Helms,AZ,85701,1953-03-20,female,white,54.0
6,7,Timothy,Allison,CT,6450,1888-04-08,male,black,60.0
7,8,Gina,McDonald,GA,30064,1984-11-27,female,white,101.0
8,9,William,Fontenot,OH,45402,1938-04-24,male,,1093.0
9,10,Maria,Howe,MO,64465,1938-12-10,female,hispanic,315.0


In [31]:
# Start at 0, finish at 100, only include every fifth row
df_patient[0:100:5]

,PatientID,FirstName,LastName,State,ZipCode,DateOfBirth,Gender,Race,Income
0,1,Loretta,Gunter,FL,32250,1979-03-29,female,white,29.0
5,6,Marlo,Helms,AZ,85701,1953-03-20,female,white,54.0
10,11,Dagmar,Brewer,NC,28405,1968-03-20,female,,73.0
15,16,Garry,Tondreau,CA,90603,1978-10-07,male,,65.0
20,21,Ruby,McAlpine,MI,48607,1995-05-23,MISSING,white,97.0
...,...,...,...,...,...,...,...,...,...
75,76,Patricia,Atkins,TX,77388,1985-06-25,female,Unknown,73.0
80,81,Rhonda,Mimms,GA,30305,1959-03-27,female,hispanic,51.0
85,86,Virgen,Hill,TX,78217,1959-04-16,female,,61.0
90,91,Carl,Pollock,NM,87102,1985-12-21,male,white,NaN


In [32]:
# last 10 rows
df_patient[-1:-11:-1]

,PatientID,FirstName,LastName,State,ZipCode,DateOfBirth,Gender,Race,Income
19999,20000,George,Pena,GA,30705,NaN,male,Missing,59.0
19998,19999,Deidre,Croft,GA,30303,1965-01-24,female,Unknown,NaN
19997,19998,Thanh,Large,FL,33912,1957-11-24,male,black,60.0
19996,19997,Wilfredo,Reinhardt,CO,80112,1950-02-23,male,other,31.0
19995,19996,Lucia,Atkins,MI,49546,1970-01-26,female,white,166.0
19994,19995,Ronald,Henry,NJ,8060,1957-05-09,male,black,54.0
19993,19994,James,Devault,OH,44902,1986-09-06,male,black,33.0
19992,19993,Raul,Seay,WA,98109,1974-04-25,male,Missing,22.0
19991,19992,Chris,Crawford,TX,78411,1939-06-30,male,other,141.0
19990,19991,Angela,Fowler,IL,61701,NaN,female,white,40.0


Requesting the first and last rows is pretty common, so there are also methods for that:

In [33]:
# First 3
df_patient.head(3)
# Last 8
df_patient.tail(8)

,PatientID,FirstName,LastName,State,ZipCode,DateOfBirth,Gender,Race,Income
0,1,Loretta,Gunter,FL,32250,1979-03-29,female,white,29.0
1,2,Todd,Rea,TX,79602,1936-12-20,male,Missing,NaN
2,3,Margaret,Goodwin,PA,18106,1948-04-19,female,hispanic,53.0


,PatientID,FirstName,LastName,State,ZipCode,DateOfBirth,Gender,Race,Income
19992,19993,Raul,Seay,WA,98109,1974-04-25,male,Missing,22.0
19993,19994,James,Devault,OH,44902,1986-09-06,male,black,33.0
19994,19995,Ronald,Henry,NJ,8060,1957-05-09,male,black,54.0
19995,19996,Lucia,Atkins,MI,49546,1970-01-26,female,white,166.0
19996,19997,Wilfredo,Reinhardt,CO,80112,1950-02-23,male,other,31.0
19997,19998,Thanh,Large,FL,33912,1957-11-24,male,black,60.0
19998,19999,Deidre,Croft,GA,30303,1965-01-24,female,Unknown,NaN
19999,20000,George,Pena,GA,30705,NaN,male,Missing,59.0


## General index methods

Pandas used to provide three methods for indexing: one (`.ix`) is now deprecated.  

The remaining two are:

* `.loc[]`    index by label location (end inclusive)
* `.iloc[]`   index by integer location, similar to numpy (end exclusive)

`.loc[]` is the primary method

They both work by passing in the row condition, then the column condition:

`.loc[rows,columns]`  
`.iloc[rows,columns]`

In [34]:
# select by label only
df_patient.loc[0:15,
               "PatientID":"LastName"]
# end inclusive

,PatientID,FirstName,LastName
0,1,Loretta,Gunter
1,2,Todd,Rea
2,3,Margaret,Goodwin
3,4,Anna,McCullough
4,5,Glenn,Labrecque
...,...,...,...
11,12,Thomas,Dvorak
12,13,Thomas,Gonzalez
13,14,Jason,Nabors
14,15,Katherine,Tanaka


In [35]:
# select by label only
df_patient.loc[0:15,
               ["PatientID","Income"]]
# end inclusive

,PatientID,Income
0,1,29.0
1,2,NaN
2,3,53.0
3,4,20.0
4,5,48.0
...,...,...
11,12,22.0
12,13,91.0
13,14,187.0
14,15,16.0


> **Take note: When slicing by `.loc` both the start label and the end label are included!**

You can pass in a Boolean condition here as well:

In [36]:
# select by label only
df_patient.loc[df_patient['Income'] > 500,
               ["PatientID","Income"]]
# end inclusive

,PatientID,Income
8,9,1093.0
60,61,1138.0
61,62,1093.0
72,73,1131.0
99,100,1079.0
...,...,...
19889,19890,1013.0
19900,19901,1063.0
19937,19938,1108.0
19952,19953,1023.0


When performing complex boolean slicing, pandas employs the `|` operator for `or`, and the `&` operator for `and`  
- surround each condition in parentheses.

In [37]:
# Multiple Boolean condition
df_patient.loc[(df_patient['Income'] > 500) &
               (df_patient['Gender'] == 'male'),
               ["PatientID","Income"]]
# end inclusive

,PatientID,Income
8,9,1093.0
61,62,1093.0
121,122,1140.0
136,137,1225.0
149,150,1024.0
...,...,...
19632,19633,1051.0
19736,19737,1196.0
19889,19890,1013.0
19900,19901,1063.0


You can also use `.iloc()` to select by integer location. This is less common in practice.

In [38]:
# select by integer location
df_patient.iloc[0:3,0:4] #not end inclusive

,PatientID,FirstName,LastName,State
0,1,Loretta,Gunter,FL
1,2,Todd,Rea,TX
2,3,Margaret,Goodwin,PA


## Importing SAS and SPSS files

Pandas has the capability to easily import SAS and SPSS files as well. There is a dependency on the module `pyreadstat` which you can install from conda. This is not part of the standard Anaconda distribution by default. The read_sas function works without it, but the SPSS reader requires it. 


In [39]:
# Sample SAS file read
sales_df = pd.read_sas(r'Data/productsales.sas7bdat')
sales_df.head()

,ACTUAL,PREDICT,COUNTRY,REGION,DIVISION,PRODTYPE,PRODUCT,QUARTER,YEAR,MONTH
0,925.0,850.0,b'CANADA',b'EAST',b'EDUCATION',b'FURNITURE',b'SOFA',1.0,1993.0,1993-01-01
1,999.0,297.0,b'CANADA',b'EAST',b'EDUCATION',b'FURNITURE',b'SOFA',1.0,1993.0,1993-02-01
2,608.0,846.0,b'CANADA',b'EAST',b'EDUCATION',b'FURNITURE',b'SOFA',1.0,1993.0,1993-03-01
3,642.0,533.0,b'CANADA',b'EAST',b'EDUCATION',b'FURNITURE',b'SOFA',2.0,1993.0,1993-04-01
4,656.0,646.0,b'CANADA',b'EAST',b'EDUCATION',b'FURNITURE',b'SOFA',2.0,1993.0,1993-05-01


In [40]:
# this requires you to conda install pyreadstat
# depress = pd.read_spss(r'Data/depress.sav')

## Review

The main points we covered in this section are:
*  Import data with Pandas
*  Understand Dataframes and Series object types
*  Basic data exploration
*  Subset and manipulate dataframes